browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+  if (it == local_storage_namespace_map().end()) {
+    auto local_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, local_partition_id);
+    local_storage_namespace_map()[local_partition_id] =
+        std::move(local_storage_namespace);
+  }
+
+  std::string session_partition_id =
+      content::GetSessionStorageNamespaceId(web_contents()) +
+      "/ephemeral-session-storage";
+  it = session_storage_namespace_map().find(session_partition_id);
+  if (it == session_storage_namespace_map().end()) {
+    auto session_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, session_partition_id);
+    session_storage_namespace_map()[session_partition_id] =
+        std::move(session_storage_namespace);
+  }
+}
+
+void EphemeralStorageTabHelper::WebContentsDestroyed() {
+  ClearEphemeralStorageIfNecessary();
+}
+
+bool EphemeralStorageTabHelper::IsDifferentWebContentsWithMatchingStorageDomain(
+    WebContents* other_web_contents,
+    const std::string& storage_domain) {
+  DCHECK(other_web_contents);
+  if (web_contents() == other_web_contents)
+    return false;
+
+  std::string other_storage_domain =
+      URLToStorageDomain(other_web_contents->GetLastCommittedURL());
+  return storage_domain == other_storage_domain;
+}
+
+bool EphemeralStorageTabHelper::IsAnotherTabOpenWithStorageDomain(
+    const std::string& storage_domain) {
+#if defined(OS_ANDROID)
+  for (auto it = TabModelList::begin(); it != TabModelList::end(); ++it) {
+    TabModel* tab_model = *it;
+    if (tab_model->GetProfile() == profile) {
+      for (int i = 0; i < tab_model->GetTabCount(); i++) {
+        TabAndroid* tab = tab_model->GetTabAt(i);
+        WebContents* web_contents = tab->web_contents();
+        if (IsDifferentWebContentsWithMatchingStorageDomain(web_contents,
+                                                            storage_domain))
+          return true;
+      }
+    }
+  }
+#else
+  for (Browser* browser : *BrowserList::GetInstance()) {
+    if (browser->profile() == web_contents()->GetBrowserContext()) {
+      TabStripModel* tab_strip = browser->tab_strip_model();
+      for (int i = 0; i < tab_strip->count(); ++i) {
+        WebContents* web_contents = tab_strip->GetWebContentsAt(i);
+        if (IsDifferentWebContentsWithMatchingStorageDomain(web_contents,
+                                                            storage_domain))
+          return true;
+      }
+    }
+  }
+#endif
+
+  return false;
+}
+
+void EphemeralStorageTabHelper::ClearEphemeralStorageIfNecessary() {
+  if (!base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage)) {
+    return;
+  }
+
+  std::string previous_domain =
+      URLToStorageDomain(web_contents()->GetLastCommittedURL());
+  if (!IsAnotherTabOpenWithStorageDomain(previous_domain)) {
+    local_storage_namespace_map().erase(previous_domain +
+                                        "/ephemeral-local-storage");
```

Comment:
can we please move this `domain + "/ephemeral-local-storage"` to a helper function

Reply:
:+1: